### ISSUE TITLE
Usage of `FromStr` vs `TryFrom<&str>`

### ISSUE THREAD

**Issue by tcharding** [rust-bitcoin/rust-bitcoin#1802]
Is this correct
> Blocks implements [FromStr](https://doc.rust-lang.org/stable/std/str/trait.FromStr.html), but not [TryFrom<&str>](https://doc.rust-lang.org/stable/std/convert/trait.TryFrom.html). This is actually intentional, because despite being nearly identical traits signature-wise they mean two very different things: FromStr implies parsing from a string whereas TryFrom<&str> is for when your data type is a subset of all strings. In our case, FromStr is the correct one to use.

(from @apoelstra's recently linked blog post: https://sabrinajewson.org/blog/errors

Does this mean that `Network` should implement `TryFrom<&str>` not `FromStr` (using `Network` as an example because I just looked at that code this morning during review).

**Comment by apoelstra** [comment]
No, `TryFrom<&str>` would imply that a `Network` somehow is a kind of string, which (in my view) it is not.

**Comment by Kixunil** [comment]
I disagree there is a semantic difference between them. `FromStr` is from the time when `TryFrom` didn't exist and I even remember discussions about `impl<'a, T: FromStr> TryFrom<&'a str> for T` and deprecating `FromStr` but it didn't happen because it would conflict with `impl<U, T: Into<U>> TryFrom<T> for U` which was more important.

I'm pretty sure if we could go back in time most developers would agree to not have `FromStr` in the first place and only use `TryFrom<&str>`.

From practical standpoint, `TryFrom<&str>` is more flexible since it *allows* borrowing the input while `FromStr` allow using the `parse()` method on `str`. Having also `TryFrom<{String, Box<str>, Rc<str>, Arc<str>}> can enable some optimal generic code. (I would've used them in [`parse_arg`](https://docs.rs/parse_arg/latest/parse_arg/) but at the time of creating some important impls were missing.) Therefore my standpoint is to implement all of them in *libraries*.

**Comment by dpc** [comment]
My understanding is the same as @Kixunil . Also

> FromStr implies parsing from a string whereas TryFrom<&str> is for when your data type is a subset of all strings. In our case, FromStr is the correct one to use.

Is so subtle, that I have to grab another coffee, and even then I'm afraid I'll never be able to correctly use it in practice. 

I'm not against re-purposing now redundant `FromStr` , but that's the first time I'm seeing someone trying to do that, and I doubt it's a well spread knowledge, so I think it will just confuse people.

**Comment by tcharding** [comment]
I'd call that consensus! Any type that has a `FromStr` should have `TryFrom<{&str, String, Box<str>, Rc<str>, Arc<str>}>`.

**Comment by tcharding** [comment]
Possibly we just add a macro that produces this:
```rust

impl TryFrom<&str> for WitnessVersion {
    type Error = ConversionError;

    fn try_from(s: &str) -> Result<Self, Self::Error> { Self::from_str(s) }
}

impl TryFrom<Box<str>> for WitnessVersion {
    type Error = ConversionError;

    fn try_from(s: Box<str>) -> Result<Self, Self::Error> { Self::from_str(&s) }
}

impl TryFrom<Arc<str>> for WitnessVersion {
    type Error = ConversionError;

    fn try_from(s: Arc<str>) -> Result<Self, Self::Error> { Self::from_str(&s) }
}

impl TryFrom<Rc<str>> for WitnessVersion {
    type Error = ConversionError;

    fn try_from(s: Rc<str>) -> Result<Self, Self::Error> { Self::from_str(&s) }
}
```

**Comment by Kixunil** [comment]
It's a bit better to call a `fn internal_parse<S: AsRef<str> + Into<String>>(string: S) -> Result<Self, E>` since we can avoid allocations.

**Comment by tcharding** [comment]
The `Into<String>` messes with the `Arc<str>` and `Rc<str>` impls, you suggest that just for the error path, right? If that is correct I don't think we need it because `std::num::IntErrorKind` has good enough error handling. Or am I missing something? This builds:
```rust
impl WitnessVersion {
    fn internal_parse<S: AsRef<str>>(string: S) -> Result<Self, ConversionError> {
        let version = string.as_ref().parse::<u8>()?;
        WitnessVersion::try_from(version)
    }
}

impl FromStr for WitnessVersion {
    type Err = ConversionError;

    fn from_str(s: &str) -> Result<Self, Self::Err> { WitnessVersion::internal_parse(s) }
}

impl TryFrom<&str> for WitnessVersion {
    type Error = ConversionError;

    fn try_from(s: &str) -> Result<Self, Self::Error> { WitnessVersion::internal_parse(s) }
}

impl TryFrom<Box<str>> for WitnessVersion {
    type Error = ConversionError;

    fn try_from(s: Box<str>) -> Result<Self, Self::Error> { WitnessVersion::internal_parse(s) }
}

impl TryFrom<Arc<str>> for WitnessVersion {
    type Error = ConversionError;

    fn try_from(s: Arc<str>) -> Result<Self, Self::Error> { WitnessVersion::internal_parse(s) }
}

impl TryFrom<Rc<str>> for WitnessVersion {
    type Error = ConversionError;

    fn try_from(s: Rc<str>) -> Result<Self, Self::Error> { WitnessVersion::internal_parse(s) }
}
```

**Comment by tcharding** [comment]
I can't see the allocation in the initial code I posted either (the body of `internal_parse` is what I had in the `FromStr` impl). What am I missing?

**Comment by stevenroose** [comment]
> I'd call that consensus! Any type that has a `FromStr` should have `TryFrom<{&str, String, Box<str>, Rc<str>, Arc<str>}>`.

After one day of conversation? I thought (rough) consensus meant that (roughly) all stakeholders agreed.

I don't want to be known for being the pedant of the group (maybe I already am), but I kinda agree with the original post that there is a subtle semantic difference. I see `FromStr` and `Display` as both making sense for types that have "one or more possible representations as a string", while `[Try]From<str>` and `Into<str>` as for types that have a "single canonical representation as a string" (kinda meaning they can be thought of as a string themselves in some way, like the author of that post suggests). The `std::convert::TryFrom` docs speak specifically of "conversion" which suggests invertibility. Like, `Into<str>` and `TryFrom<str>` would make sense for `Address`, whereas it wouldn't for `Amount`. I agree that any type that has `Into<str>` and `TryFrom<str>` should obviously also have proxying impls for `FromStr` and `Display`, I wouldn't argue the reverse.

In fact, to add to @Kixunil's argument about `TryFrom` allowing for a more efficient impl by not needing to potentially re-allocate the string, that argument actually kinda follows the same argument because it's only possible if the type actually *is* the string or is gonna use the string to be part of itself. In the types where along the above argument `TryFrom` and `Into` would not make sense, you will probably never be able to use the `String` in the conversion. (Unless you would maybe to fancy substring-ing without re-allocation or something but afaik Rust doesn't have APIs for that.)

**Comment by Kixunil** [comment]
@tcharding for `Arc` and `Rc` we need to pass `&*s`. They aren't costlessly convertible anyway. (Well, I have some crazier ideas but not sure if wee should do it.)

@stevenroose not entirely, if we want to capture the string in error then `TryFrom<String>` is costless on error path. But my larger point was about being used in generic code. Being able to treat both borrowed and non-borrowed strings the same is just nicer.

Anyway, this debate may belong to URLO since we want to be in sync with Rust ecosystem.

**Comment by stevenroose** [comment]
> Anyway, this debate may belong to URLO since we want to be in sync with Rust ecosystem.

You mean like updating MSRV as soon as new stable is out? ;)

Fwiw, personally I don't care about error case optimization 99.9% of the time. Few users will be doing thousands of iterations of an error case.

**Comment by Kixunil** [comment]
Heh, I half-expected this kind of snark but I hoped everyone understands that following Rust guidelines etc is recommendation, not obligation.

I personally like to avoid allocations or other if I can without significantly degrading other properties of the code. In this case I see no downside except for supposed semantic difference which I don't believe in.

**Comment by dpc** [comment]
> I don't want to be known for being the pedant of the group (maybe I already am), but I kinda agree with the original post that there is a subtle semantic difference.

It's good to be a pedant when working on an important library like rust-bitcoin.

@Kixunil Seems like initial couple of voices are pro-semantic-difference. Looking at it a little bit more - if there was supposed to be no difference, why is standard library not implementing `TryFrom<String> for i32`, again? Were there any technical reasons or are we waiting for something?

I think we can ignore how these types came to be. It's not very important. If the community thinks it's worthwhile to have a difference, and we already have the traits, so there can be a difference.

OK... So `FromStr` is for parsing from `&'str` without ability to take over the ownership (because it's parsing, not a conversion)...

... since I'm a conformist, let me look at the past discussion from that PoV...

> No, TryFrom<&str> would imply that a Network somehow is a kind of string, which (in my view) it is not.


OK. So here I would dissagree. Is it just "network" in some abstract sense or rather "network designator" or "network type". Because if you have a `"mainnet"` or `"bitcoin"` string it very much *is* a designator of an network. It doesn't even require any "parsing" (like checking digits). It's a straight lookup in a list.

So I'd say that even assuming `FromStr` is distinct, `Network` should still `impl TryFrom<String>` etc.

Looking more practically - is there any benefit (class of bugs prevented) that would make not implementing it better than implementing it for the consumers of this library?

**Comment by apoelstra** [comment]
> So I'd say that even assuming `FromStr` is distinct, `Network` should still `impl TryFrom<String>` etc.

Ok, I buy this.

**Comment by tcharding** [comment]
> After one day of conversation? I thought (rough) consensus meant that (roughly) all stakeholders agreed.

Yes that's what I meant :) We had Andrew, dpc, and Kix all in agreement, that is one more than roughly all stakeholders that show up on a daily basis over the last 18 months. This topic is mainly of academic interest so there is no risk to getting it wrong. It hasn't sunk in yet that you are showing up regularly but it will soon, no doubt!

> I don't want to be known for being the pedant of the group

You'll have to fight for that title, we have a few proud contenders. Legendary usage of language BTW, I had to look up what "pedant" means - such a good description for how I often feel.


As for the rest of the discussion, I'm struggling with the difference between "parsing" and int from "6" and "conversion" of an int for "6"? I also struggled with the whole "when to implement `From<Foo>` and when to use inherent `from_foo` discussion. Is the string "6" _a number_ or is it a string representation of a number that must be parsed, I'm not sure I get the subtlety?

**Comment by Kixunil** [comment]
> why is standard library not implementing `TryFrom<String> for i32`? Were there any technical reasons or are we waiting for something?

I strongly believe that the reason is "nobody made a PR for it". I'm actually thinking of making it though I worry it might border on abuse of ACP.

> Looking more practically - is there any benefit (class of bugs prevented) that would make not implementing it better than implementing it for the consumers of this library?

Exactly, I don't see any. If someone provides a real example of harm I will be first to drop this.

> I'm struggling with the difference between "parsing" and int from "6" and "conversion" of an int for "6"?

Me too and I guess it's because there's no meaningful difference between the other than "one makes more instructions than the other during conversion by some arbitrary margin".

**Comment by dpc** [comment]
> Me too and I guess it's because there's no meaningful difference between the other than "one makes more instructions than the other during conversion by some arbitrary margin".

To me it's: if you need to algorithmically calculate a different representation from a string encoding, then it's parsing. If it's near to 1:1 match / lookup, then it passes as a conversion. I agree that it's fuzzy and somewhat arbitrary, but it doesn't invalidate it necessarily.

**Comment by stevenroose** [comment]
> It hasn't sunk in yet that you are showing up regularly but it will soon, no doubt!

I feel that, haha. And don't want to set expectations either.. But anyway.

> I had to look up what "pedant" means 

"someone who has the habit of being pedantic"?

To me it seems like every reply in the forums agrees there is a semantic difference. But then again people lingering forums are usually the more pedantic kind :sweat_smile:  

To me the semantic difference is equivalent to the semantic difference between parsing and converting. I don't know if that helps. Conversion more strongly implies there is an expected canonical string. But I wouldn't say that's all. Yeah it's hard to pin down, but to me there is a semantic difference between parsing and converting. Like a conversion you expect to "generally work, except in some cases when there was some error". While parsing is more like "let's see if this in any way makes sense to be of this type". 

Also, tbh I think the implementation stdlib has for both TryFrom<string> (none) and FromStr. All of those types I would expect have TryFrom<string> (Paths etc) actually have From<string>.

Think like this: `5u64` and `5u32` are obviously exactly the same semantic value of five, but need a `TryFrom` conversion because obviously there are some u64's that are not u32's.

**Comment by rockcoolsaint** [comment]
I'm picking this up to work on it

**Comment by tcharding** [comment]
We don't assign, just go wild and push up a PR. Thanks!

**Comment by rockcoolsaint** [comment]
> We don't assign, just go wild and push up a PR. Thanks!

Alright, thanks for the feedback

**Comment by Sh0g0-1758** [comment]
> The `Into<String>` messes with the `Arc<str>` and `Rc<str>` impls, you suggest that just for the error path, right? If that is correct I don't think we need it because `std::num::IntErrorKind` has good enough error handling. Or am I missing something? This builds:
> 
> ```rust
> impl WitnessVersion {
>     fn internal_parse<S: AsRef<str>>(string: S) -> Result<Self, ConversionError> {
>         let version = string.as_ref().parse::<u8>()?;
>         WitnessVersion::try_from(version)
>     }
> }
> 
> impl FromStr for WitnessVersion {
>     type Err = ConversionError;
> 
>     fn from_str(s: &str) -> Result<Self, Self::Err> { WitnessVersion::internal_parse(s) }
> }
> 
> impl TryFrom<&str> for WitnessVersion {
>     type Error = ConversionError;
> 
>     fn try_from(s: &str) -> Result<Self, Self::Error> { WitnessVersion::internal_parse(s) }
> }
> 
> impl TryFrom<Box<str>> for WitnessVersion {
>     type Error = ConversionError;
> 
>     fn try_from(s: Box<str>) -> Result<Self, Self::Error> { WitnessVersion::internal_parse(s) }
> }
> 
> impl TryFrom<Arc<str>> for WitnessVersion {
>     type Error = ConversionError;
> 
>     fn try_from(s: Arc<str>) -> Result<Self, Self::Error> { WitnessVersion::internal_parse(s) }
> }
> 
> impl TryFrom<Rc<str>> for WitnessVersion {
>     type Error = ConversionError;
> 
>     fn try_from(s: Rc<str>) -> Result<Self, Self::Error> { WitnessVersion::internal_parse(s) }
> }
> ```

Hey @tcharding , needed a little sanity check. Do these changes need to be implemented only for `WitnessVersion` ? and in the implementation, you refer to `ConversionError` which is not in the original file. So should I make a custom enum for the implementation?

**Comment by tcharding** [comment]
Its in the title :) "Types that implement FromStr should implement TryFrom<&str> and friends)"

But I don't know the validity of this issue, I started reading the thread again and the linked discussion on URLO but I don't know what others want to do about this.

**Comment by apoelstra** [comment]
I continue to dislike this idea and think that `From` is about converting between string-like types and `FromStr` is about parsing strings. But I won't hold it up because in the linked issue apparently many other people disagree and have some idea about what they want done here.

**Comment by Kixunil** [comment]
@apoelstra if there wasn't potential performance optimization I wouldn't want to do it either. Maybe we could have `parse_arg::ParseArg` instead which is more general and achieves the same thing but it could look a bit weird to use that to parse non-arg strings.

**Comment by apoelstra** [comment]
I think we should do the most idiomatic thing, which is `FromStr`.

If we also have `From<&str>` I think that's mostly harmless, since nobody will realize that it exists.

**Comment by Kixunil** [comment]
I never said to drop `FromStr`...

The real power of these generics comes from having both `TryFrom<&str>` and `TryFrom<String>`.
